---
title: "README"
author: "Rasmus Kirkegaard"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(scipen = 10)
```

# R10.4.1 Zymo HMW basecalling

With the release of R10.4.1 I wanted to check the quality of the raw reads and the consensus accuracy to ensure that we had the right approach for continuing our nanopore only assemblies from metagenomic samples ([Mantas et al. 2022](https://www.nature.com/articles/s41592-022-01539-7)). So we sequenced the Zymo HMW mock DNA to evaluate the quality of the raw reads but more importantly the assembled genomes. With the introduction of [dorado](https://github.com/nanoporetech/dorado) as a basecaller that should replace guppy it was also interesting to see how well that performed.

## Aim

1. Check if nanopore R10.4.1 is also capable of generating nanopore only assemblies. 
2. Evaluate the impact is of the different basecalling models on raw read accuracy 
3. Evaluate the impact is of the different basecalling models on consensus accuracy

## Conclusion

1. Yes!!! The obtained indel and mismatch rates are very low already around 25X coverage. At ~0.1 /100 Kbp a 5 Mbp genome will have 5 "errors" (Could still be the references at this point but likely does not matter) so it is unlikely to affect gene calling and most downstream analysis.
2. There is a huge difference between fast, hac and sup for raw read accuracy. However, the time needed for compute is also much much higher for the super accuracy model.
3. There seems to be a clear improvement in consensus accuracy by using super accuracy reads. Even providing higher coverage for fast or hac accuracy reads do not seem to fix this indicating that some of the errors fixed by sup(er accuracy) are indeed systematic errors in fast and hac. So super accuracy mode is the way to go if the aim is to generate reference quality genomes despite the additional need for GPU compute.

## Data availability

Uploading >1TB fast5 files have turned out to be a complicated challenge... seems like most of the guides for this were written in the early MinION access program days and never updated as they still refer to [fast5 files with basecalls included, albacore, and metrichor](https://ena-docs.readthedocs.io/en/latest/submit/fileprep/reads.html#oxford-nanopore). If someone has a good solution please let me know and I will be happy to deposit the data.

```{r load_libraries_and_data,echo=FALSE,warning=FALSE,message=FALSE}
library(tidyverse)
library(gridExtra)
tt2 <- ttheme_default(core=list(fg_params=list(hjust=1, x = 0.95, fontsize = 10)),
                      colhead=list(fg_params=list(fontsize = 12)))

# duplex read QC new ref
get_read_QC<-function(model="dna_r10.4.1_e8.2_400bps_hac@v4.0.0") {
  files<-dir(pattern = paste0(model,"_ref_E.coli_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Escherichia_coli <- files %>% map(read_tsv) %>% reduce(rbind) %>% mutate(REF="Escherichia_coli")
  files<-dir(pattern = paste0(model,"_ref_L.monocytogenes_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Listeria_monocytogenes<-files %>% map(read_tsv) %>% reduce(rbind) %>% mutate(REF="Listeria_monocytogenes")
  files<-dir(pattern = paste0(model,"_ref_P.aeruginosa_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Pseudomonas_aeruginosa<- files %>% map(read_tsv) %>% reduce(rbind) %>% mutate(REF="Pseudomonas_aeruginosa")
  files<-dir(pattern = paste0(model,"_ref_B.subtilis_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Bacillus_subtilis<- files %>% map(read_tsv) %>% reduce(rbind) %>% mutate(REF="Bacillus_subtilis")
  files<-dir(pattern = paste0(model,"_ref_S.enterica_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Salmonella_enterica<-files %>% map(read_tsv) %>% reduce(rbind) %>% mutate(REF="Salmonella_enterica")
  files<-dir(pattern = paste0(model,"_ref_E.faecalis_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Enterococcus_faecalis<-files %>% map(read_tsv) %>% reduce(rbind) %>% mutate(REF="Enterococcus_faecalis")
  files<-dir(pattern = paste0(model,"_ref_S.aureus_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Staphylococcus_aureus<-files %>% map(read_tsv) %>% reduce(rbind) %>% mutate(REF="Staphylococcus_aureus")
  NP_QC_combined<-bind_rows(NP_QC_ref_Bacillus_subtilis,
                            NP_QC_ref_Enterococcus_faecalis,
                            NP_QC_ref_Escherichia_coli,
                            NP_QC_ref_Listeria_monocytogenes,
                            NP_QC_ref_Pseudomonas_aeruginosa,
                            NP_QC_ref_Salmonella_enterica,
                            NP_QC_ref_Staphylococcus_aureus) %>% 
    mutate(model=model)
return(NP_QC_combined)
}
NP_QC_fast<-get_read_QC(model = "dna_r10.4.1_e8.2_400bps_fast@v4.0.0")
NP_QC_hac<-get_read_QC(model = "dna_r10.4.1_e8.2_400bps_hac@v4.0.0")
NP_QC_sup<-get_read_QC(model = "dna_r10.4.1_e8.2_400bps_sup@v4.0.0")
NP_QC_combined<-bind_rows(NP_QC_fast,NP_QC_hac,NP_QC_sup)

### read coverage
draft_assemblies<-list.files(path = "results/",pattern = "*.assembly_info.txt",full.names = T)
cov_tab<-tibble()
for (f in draft_assemblies) {
  cov_temp<-read_tsv(file = f) %>% mutate(draft=f)
  cov_tab<-bind_rows(cov_tab,cov_temp)
}
cov_tab<-cov_tab %>% mutate(contigName=`#seq_name`,coverage=cov.,circular=circ.,
                            ReadSet=str_replace(string =draft,pattern = ".*(dna.*).assembly_info.txt",replacement = "\\1")) %>% 
  select(ReadSet,contigName,coverage,circular)

### Assembly based stats
fastani<-read_tsv(file = "results/fastani.tsv",col_names = c("query","ref","ANI","aligned_segments","total_segments")) %>% 
  mutate(Assembly=str_replace(string = query,pattern = "results/bins/(.*).fa.*",replacement = "\\1"),
         ref=str_replace(string = ref,pattern = ".*/(.*).fasta",replacement = "\\1")) %>% 
  select(Assembly,ref,ANI)

ref_genomes=list.files(path = "results/",pattern = "quast_.*.tsv",full.names = T)

# Load quast data
quast_tab<-tibble()
for (f in ref_genomes) {
  quast_temp<-read_tsv(file = f) %>% mutate(ref=f) %>% filter(`Total aligned length`!="-")
  quast_tab<-bind_rows(quast_tab,quast_temp)
}
quast_tab<-quast_tab %>% mutate(ref=str_replace(string = ref,pattern = ".*quast_(.*).tsv",replacement = "\\1"),
                                Assembly=str_trim(Assembly)) %>% filter(as.numeric(`Total aligned length`)>10^6) %>%
  select(Assembly,ref,`Total length`,`Reference length`,`Largest alignment`,`Total aligned length`,`# indels per 100 kbp`,`# mismatches per 100 kbp`) %>%
  mutate(`Total aligned length`=as.numeric(`Total aligned length`),
         `# indels per 100 kbp`=as.numeric(`# indels per 100 kbp`),
         `# mismatches per 100 kbp`=as.numeric(`# mismatches per 100 kbp`)) %>%
  filter(Assembly!=ref)

###
genome_stats<-fastani %>% right_join(quast_tab,by = c("Assembly", "ref")) %>%
  mutate(contigName=str_replace(string = Assembly,pattern = ".*(contig.*)",replacement = "\\1"),
         ReadSet=str_replace(string = Assembly,pattern = "(.*).flye.*",replacement = "\\1")) %>%
  filter(ANI>95) %>%
  left_join(cov_tab,by = c("contigName", "ReadSet")) %>%
  mutate(asmtype=str_replace(string = Assembly,pattern = ".*(fly.*).contig.*",replacement = "\\1")) %>%
  mutate(model=str_replace(ReadSet,pattern = "-.*",replacement = "")) %>%
  mutate(ref=str_replace(ref,pattern = "_hifiasm",replacement = "")) %>%
  filter(ref!="S.cerevisiae")
```

## NP reads aligned to the refs

```{r fig.width=10,fig.height=7,echo=FALSE,warning=FALSE,message=FALSE}
NP_QC_combined %>% ggplot(aes(x = percentIdentity,col=REF))+geom_density()+geom_vline(xintercept = 99,col="red",linetype="dashed")+facet_wrap(~model,ncol = 1)+scale_x_continuous(breaks = 80:100,limits = c(80,100))
```

## Indel rate vs coverage

```{r fig.width=10, echo=FALSE,warning=FALSE,message=FALSE}
genome_stats %>% 
  ggplot(aes(x=coverage,y = `# indels per 100 kbp`,col=model,shape=asmtype))+geom_point()+geom_line()+facet_wrap(~ref)+scale_y_log10()+coord_cartesian(xlim=c(0, 100))
```

## Mismatch rate vs coverage

```{r fig.width=10, echo=FALSE,warning=FALSE,message=FALSE}
genome_stats %>% 
  ggplot(aes(x=coverage,y = `# mismatches per 100 kbp`,col=model,shape=asmtype))+geom_point()+geom_line()+facet_wrap(~ref)+scale_y_log10()+coord_cartesian(xlim=c(0, 100))
```

## Materials and methods

Here is a brief description of the tools used. For the exact commands check out the **Snakefile** in this repository ([Snakemake](https://snakemake.readthedocs.io/en/stable/) v. 7.18.2).

### DNA sequencing
DNA sample was the [Zymo Mock HMW standard](https://zymoresearch.eu/products/zymobiomics-hmw-dna-standard). The DNA was prepared for sequencing using the nanopore ligation sequencing kit (SQK-LSK114) and sequenced on a R10.4.1 nanopore promethion flowcell (FLO-PRO114M) with the "400 bp/s" mode (4khz sampling). 

### Basecalling
The reads were basecalled using [dorado](https://github.com/nanoporetech/dorado) (v. 0.1.1) with fast, hac and sup accuracy mode using the 4.0.0 models. 

### Read QC
Reads were mapped to the updated zymo reference genomes (hopefully goes public soon) using [minimap2](https://github.com/lh3/minimap2) (v. 2.24), and QC information was obtained using [NanoPlot](https://github.com/wdecoster/NanoPlot) (v. 1.41.0).

### Assembly
The reads were subsampled using [seqtk](https://github.com/lh3/seqtk) (v. 1.3) and assembled using [flye](https://github.com/fenderglass/Flye) (v. 2.9.1). The metagenome assemblies were then polished using [medaka](https://github.com/nanoporetech/medaka) (v. 1.7.2).

### Genome quality assessment
The assembled contigs were compared to the reference contigs using [QUAST](https://github.com/ablab/quast) (v. 5.2.0) and [fastANI](https://github.com/ParBLiSS/FastANI) (v. 1.33).